File: packages/sdk-py/examples/canary_rollout.py
"""Start a canary rollout and poll until it reaches a terminal state.

Usage::

    python examples/canary_rollout.py <service_id> <digest> <percentage> <change_ticket_url>

Environment variables:
    ENCLII_TOKEN     Required. API token.
    ENCLII_API_URL   Optional. Defaults to https://api.enclii.dev.
"""

from __future__ import annotations

import asyncio
import os
import sys

from enclii_sdk import AsyncEncliiClient


async def main(
    service_id: str, digest: str, percentage: int, change_ticket_url: str
) -> int:
    async with AsyncEncliiClient(
        base_url=os.environ.get("ENCLII_API_URL", "https://api.enclii.dev"),
        token=os.environ["ENCLII_TOKEN"],
    ) as enclii:
        print(f"Starting canary: {digest} @ {percentage}%…")
        rollout = await enclii.canary.start(
            service_id,
            digest=digest,
            percentage=percentage,
            validation_window_minutes=10,
            change_ticket_url=change_ticket_url,
        )
        print(f"  rollout_id={rollout.id}")
        print(f"  initial_state={rollout.state}")
        print(f"  actual_traffic={rollout.actual_percentage}%")

        while rollout.state.is_active():
            await asyncio.sleep(15)
            rollout = await enclii.canary.get(service_id, str(rollout.id))
            print(f"  state={rollout.state}")

        print(f"\nTerminal state: {rollout.state}")
        if rollout.last_error:
            print(f"  error: {rollout.last_error}")
        if rollout.rollback_reason:
            print(f"  rollback_reason: {rollout.rollback_reason}")

        return 0 if rollout.state.value == "succeeded" else 1


if __name__ == "__main__":
    if len(sys.argv) < 5:
        print(__doc__)
        sys.exit(2)
    sys.exit(
        asyncio.run(
            main(sys.argv[1], sys.argv[2], int(sys.argv[3]), sys.argv[4])
        )
    )

File: packages/sdk-ts/examples/canary-rollout.ts
/**
 * Start a canary rollout and block until it reaches a terminal state
 * (succeeded, auto_rolled_back, manual_rolled_back, failed).
 *
 * Usage:
 *   ENCLII_TOKEN=... ENCLII_SERVICE_ID=... ENCLII_DIGEST=sha256:... \
 *     tsx examples/canary-rollout.ts
 */

import { EncliiClient, isCanaryTerminal } from '@madfam/enclii-sdk';

async function main() {
  const client = new EncliiClient({
    baseUrl: process.env.ENCLII_BASE_URL ?? 'https://api.enclii.dev/v1',
    token: process.env.ENCLII_TOKEN,
  });

  const serviceId = req('ENCLII_SERVICE_ID');
  const digest = req('ENCLII_DIGEST');

  console.log(`Starting 20% canary for ${serviceId} @ ${digest}...`);
  const rollout = await client.canary.start(serviceId, {
    digest,
    percentage: 20,
    validation_window_minutes: 10,
    error_rate_threshold: 0.01,
    change_ticket_url: process.env.ENCLII_CHANGE_TICKET,
  });
  console.log(`Canary ${rollout.id} in state ${rollout.state}.`);

  // Poll until terminal.
  for (;;) {
    await sleep(10_000);
    const r = await client.canary.get(serviceId, rollout.id);
    console.log(
      `  [${r.state}] canary=${r.canary_replicas}/${r.total_replicas} ` +
        `(~${r.actual_percentage?.toFixed(1) ?? '?'}%)`,
    );
    if (isCanaryTerminal(r.state)) {
      if (r.state === 'succeeded') {
        console.log('Canary promoted to stable.');
        process.exit(0);
      }
      console.error(
        `Canary ended in ${r.state}${r.last_error ? `: ${r.last_error}` : ''}`,
      );
      process.exit(1);
    }
  }
}

function req(name: string): string {
  const v = process.env[name];
  if (!v) {
    console.error(`missing ${name}`);
    process.exit(64);
  }
  return v;
}

function sleep(ms: number) {
  return new Promise((r) => setTimeout(r, ms));
}

main();
